diff --git a/toolkit/mozapps/extensions/internal/XPIDatabase.sys.mjs b/toolkit/mozapps/extensions/internal/XPIDatabase.sys.mjs
index aca68ee50a1e..2666c1c90b9b 100644
--- a/toolkit/mozapps/extensions/internal/XPIDatabase.sys.mjs
+++ b/toolkit/mozapps/extensions/internal/XPIDatabase.sys.mjs
@@ -3688,7 +3688,10 @@ export const XPIDatabaseReconcile = {
         xpiState,
         newAddon
       );
-    } else if (oldAddon.path != xpiState.path) {
+    } else if (
+      oldAddon.path != xpiState.path ||
+      oldAddon.rootURI != xpiState.rootURI
+    ) {
       newAddon = this.updatePath(installLocation, oldAddon, xpiState);
     } else if (aUpdateCompatibility || aSchemaChange) {
       newAddon = this.updateCompatibility(
