diff --git a/toolkit/mozapps/extensions/internal/XPIProvider.sys.mjs b/toolkit/mozapps/extensions/internal/XPIProvider.sys.mjs
index 85e3f015d29e..1d41543f057e 100644
--- a/toolkit/mozapps/extensions/internal/XPIProvider.sys.mjs
+++ b/toolkit/mozapps/extensions/internal/XPIProvider.sys.mjs
@@ -1613,6 +1613,19 @@ var XPIStates = {
           startupScanScopes |= AddonManager.SCOPE_PROFILE;
         }
       }
+
+      const currentProfileLocation = XPIStates.getLocation(KEY_APP_PROFILE);
+      const previousProfilePath = oldState[KEY_APP_PROFILE]?.path;
+      if (
+        currentProfileLocation?.path &&
+        previousProfilePath &&
+        currentProfileLocation.path !== previousProfilePath
+      ) {
+        logger.warn(
+          `Force scan SCOPE_PROFILE location (app-profile path changed from ${previousProfilePath} to ${currentProfileLocation.path})`
+        );
+        startupScanScopes |= AddonManager.SCOPE_PROFILE;
+      }
     }
 
     for (let loc of XPIStates.locations()) {
@@ -1681,9 +1694,13 @@ var XPIStates = {
         } else {
           let addonChanged = false;
           if (entry instanceof Ci.nsIFile) {
+            const currentRootURI = getURIForResourceInFile(entry, "").spec;
             addonChanged =
-              xpiState.getModTime(entry) || entry.path != xpiState.path;
+              xpiState.getModTime(entry) ||
+              entry.path != xpiState.path ||
+              currentRootURI != xpiState.rootURI;
             xpiState.file = entry.clone();
+            xpiState.rootURI = currentRootURI;
           } else {
             addonChanged =
               buildIdChanged ||
