diff --git a/widget/uikit/moz.build b/widget/uikit/moz.build
index 0ea48673ee27..2133a6aaae95 100644
--- a/widget/uikit/moz.build
+++ b/widget/uikit/moz.build
@@ -15,10 +15,14 @@ EXPORTS += [
 ]
 
 EXPORTS.mozilla.widget += [
+    "CompositorWidgetChild.h",
+    "CompositorWidgetParent.h",
     "EventDispatcher.h",
+    "GeckoViewRuntimeSupport.h",
     "GeckoViewSupport.h",
     "IOSView.h",
     "nsWindow.h",
+    "UIKitCompositorWidget.h",
 ]
 
 EXPORTS.GeckoView += [
@@ -26,12 +30,15 @@ EXPORTS.GeckoView += [
 ]
 
 SOURCES += [
+    "CompositorWidgetChild.cpp",
+    "CompositorWidgetParent.cpp",
     "EventDispatcher.mm",
     "GfxInfo.cpp",
     "IOSBridge.mm",
     "MediaKeysEventSourceFactory.cpp",
     "nsAppShell.mm",
     "nsBidiKeyboard.mm",
+    "nsClipboard.mm",
     "nsLookAndFeel.mm",
     "nsNativeThemeUIKit.mm",
     "nsUserIdleServiceUIKit.cpp",
@@ -39,6 +46,7 @@ SOURCES += [
     "nsWindow.mm",
     "ScreenHelperUIKit.mm",
     "TextInputHandler.mm",
+    "UIKitCompositorWidget.cpp",
     "UIKitUtils.mm",
 ]
 
